File: android/app/src/main/kotlin/dev/rohanjsh/pitcher/casting/CastingHostApiImpl.kt
package dev.rohanjsh.pitcher.casting

import android.content.Context
import android.util.Log

class CastingHostApiImpl(
    context: Context,
    private val flutterApi: CastingFlutterApi
) : CastingHostApi {

    private val chromecastManager = ChromecastManager(context)

    private var activeProvider: CastingProvider? = null
    private var currentMedia: MediaInfo? = null
    private var isDiscovering = false

    private val activePlaybackProvider: CastingPlaybackProvider?
        get() = when (activeProvider) {
            CastingProvider.CHROMECAST -> chromecastManager
            CastingProvider.AIRPLAY, null -> null
        }

    init {
        chromecastManager.setStateChangeHandler { state ->
            handleProviderStateChange(CastingProvider.CHROMECAST, state)
        }
        chromecastManager.setDevicesChangeHandler { devices ->
            handleDevicesChanged(devices)
        }
    }

    override fun startDiscovery() {
        if (isDiscovering) {
            log("Discovery already active")
            return
        }

        log("Starting discovery...")
        isDiscovering = true
        chromecastManager.startDiscovery()
    }

    override fun stopDiscovery() {
        if (!isDiscovering) return

        log("Stopping discovery...")
        isDiscovering = false
        chromecastManager.stopDiscovery()
    }

    override fun getDiscoveredDevices(): List<CastDevice> {
        return chromecastManager.getDiscoveredDevices()
    }

    override fun connect(deviceId: String) {
        val chromecastDevice = chromecastManager.getDiscoveredDevices().find { it.id == deviceId }

        when {
            chromecastDevice != null -> {
                log("Connecting to Chromecast: ${chromecastDevice.name}")
                activeProvider = CastingProvider.CHROMECAST
                chromecastManager.connect(deviceId)
            }
            else -> {
                log("Device not found: $deviceId")
                notifyState(
                    connectionState = CastingConnectionState.DISCONNECTED,
                    playbackState = CastingPlaybackState.ERROR,
                    errorMessage = "Device not found: $deviceId"
                )
            }
        }
    }

    override fun disconnect() { activePlaybackProvider?.disconnect() ?: log("No active session to disconnect"); activeProvider = null }
    override fun showAirPlayPicker() { log("showAirPlayPicker called on Android (no-op)") }

    override fun loadMedia(mediaInfo: MediaInfo, autoplay: Boolean, positionMs: Long) {
        currentMedia = mediaInfo; activePlaybackProvider?.loadMedia(mediaInfo, autoplay, positionMs) ?: log("Cannot load media: no active session")
    }
    override fun play() { activePlaybackProvider?.play() }
    override fun pause() { activePlaybackProvider?.pause() }
    override fun seek(positionMs: Long) { activePlaybackProvider?.seek(positionMs) }
    override fun stop() { activePlaybackProvider?.stop(); currentMedia = null }
    override fun setVolume(volume: Double) { activePlaybackProvider?.setVolume(volume) }

    override fun setMuted(muted: Boolean) {
        when (activeProvider) {
            CastingProvider.CHROMECAST -> chromecastManager.setMuted(muted)
            CastingProvider.AIRPLAY, null -> log("No active session for setMuted")
        }
    }

    fun dispose() { log("Disposing..."); stopDiscovery(); chromecastManager.dispose() }

    private fun handleProviderStateChange(provider: CastingProvider, state: ProviderState) {
        when (state.connectionState) {
            CastingConnectionState.CONNECTED, CastingConnectionState.CONNECTING -> activeProvider = provider
            CastingConnectionState.DISCONNECTED -> if (activeProvider == provider) { activeProvider = null; currentMedia = null }
        }
        if (activeProvider != provider && activeProvider != null) return
        notifyState(state.connectionState, state.playbackState, state.device, state.positionMs.takeIf { it > 0 }, state.durationMs.takeIf { it > 0 }, state.error)
    }

    private fun handleDevicesChanged(devices: List<CastDevice>) {
        log("Devices changed: ${getDiscoveredDevices().size} total")
        flutterApi.onDevicesChanged(getDiscoveredDevices()) { }
    }

    private fun notifyState(connectionState: CastingConnectionState, playbackState: CastingPlaybackState,
                            connectedDevice: CastDevice? = null, positionMs: Long? = null, durationMs: Long? = null, errorMessage: String? = null) {
        flutterApi.onStateChanged(CastingState(connectionState, playbackState, connectedDevice, currentMedia, positionMs, durationMs, errorMessage)) { }
    }

    private fun log(message: String) = Log.d(TAG, message)
    companion object { private const val TAG = "CastingHostApi" }
}

File: ios/Runner/Casting/CastingHostApiImpl.swift
import Foundation
import GoogleCast
import os.log

@MainActor
final class CastingHostApiImpl: NSObject, @preconcurrency CastingHostApi {

    private let flutterApi: CastingFlutterApi
    private let chromecastManager: ChromecastManager
    private let airPlayManager: AirPlayManager

    private var isDiscovering = false
    private var currentMedia: MediaInfo?
    private var activeProvider: CastingProvider?

    private var activePlaybackProvider: CastingPlaybackProvider? {
        switch activeProvider {
        case .chromecast: return chromecastManager
        case .airplay: return airPlayManager
        case nil: return nil
        }
    }

    init(flutterApi: CastingFlutterApi, castContext: GCKCastContext) {
        self.flutterApi = flutterApi
        self.chromecastManager = ChromecastManager(castContext: castContext)
        self.airPlayManager = AirPlayManager()
        super.init()

        setupManagerCallbacks()
        log("Initialized (Chromecast + AirPlay)")
    }

    deinit {
        dispose()
    }

    private func setupManagerCallbacks() {
        chromecastManager.setStateChangeHandler { [weak self] state in
            Task { @MainActor in self?.handleProviderStateChange(state, provider: .chromecast) }
        }
        chromecastManager.setDevicesChangeHandler { [weak self] _ in
            Task { @MainActor in self?.refreshDiscoveredDevices() }
        }
        airPlayManager.setStateChangeHandler { [weak self] state in
            Task { @MainActor in self?.handleProviderStateChange(state, provider: .airplay) }
        }
    }

    func startDiscovery() throws {
        guard !isDiscovering else {
            log("Discovery already active")
            return
        }

        log("Starting discovery...")
        isDiscovering = true
        chromecastManager.startDiscovery()
        airPlayManager.startRouteDetection()
        refreshDiscoveredDevices()
    }

    func stopDiscovery() throws {
        guard isDiscovering else { return }

        log("Stopping discovery...")
        isDiscovering = false
        chromecastManager.stopDiscovery()
        airPlayManager.stopRouteDetection()
    }

    func getDiscoveredDevices() throws -> [CastDevice] {
        return getAggregatedDevices()
    }

    func connect(deviceId: String) throws {
        if deviceId.hasPrefix("airplay_") {
            log("AirPlay selected - showing picker")
            activeProvider = .airplay
            airPlayManager.showPicker()
        } else {
            log("Connecting to Chromecast: \(deviceId)")
            activeProvider = .chromecast
            chromecastManager.connect(deviceId: deviceId)
        }
    }

    func disconnect() throws {
        activePlaybackProvider?.disconnect()
        activeProvider = nil
    }

    func showAirPlayPicker() throws {
        activeProvider = .airplay
        airPlayManager.showPicker()
    }

    func loadMedia(mediaInfo: MediaInfo, autoplay: Bool, positionMs: Int64) throws {
        currentMedia = mediaInfo
        activePlaybackProvider?.loadMedia(mediaInfo, autoplay: autoplay, positionMs: positionMs)
    }

    func play() throws {
        activePlaybackProvider?.play()
    }

    func pause() throws {
        activePlaybackProvider?.pause()
    }

    func seek(positionMs: Int64) throws {
        activePlaybackProvider?.seek(to: positionMs)
    }

    func stop() throws {
        activePlaybackProvider?.stop()
        currentMedia = nil
    }

    func setVolume(volume: Double) throws {
        activePlaybackProvider?.setVolume(Float(volume))
    }

    func setMuted(muted: Bool) throws {
        switch activeProvider {
        case .chromecast: chromecastManager.setMuted(muted)
        case .airplay: airPlayManager.setVolume(muted ? 0 : 1)
        case nil: break
        }
    }

    nonisolated func dispose() {
        Task { @MainActor in
            log("Disposing...")
            try? stopDiscovery()
            chromecastManager.stop()
            airPlayManager.stop()
        }
    }

    private func handleProviderStateChange(_ state: ProviderState, provider: CastingProvider) {
        switch state.connectionState {
        case .connected, .connecting:
            activeProvider = provider
        case .disconnected:
            if activeProvider == provider { activeProvider = nil; currentMedia = nil }
        }

        guard activeProvider == provider || activeProvider == nil else { return }

        flutterApi.onStateChanged(state: CastingState(
            connectionState: state.connectionState, playbackState: state.playbackState, connectedDevice: state.device,
            currentMedia: currentMedia, positionMs: state.positionMs, durationMs: state.durationMs, errorMessage: state.error
        )) { _ in }
    }

    private func getAggregatedDevices() -> [CastDevice] {
        var devices = chromecastManager.getDiscoveredDevices()
        if airPlayManager.multipleRoutesDetected {
            devices.append(CastDevice(id: "airplay_available", name: "AirPlay", provider: .airplay, modelName: "Tap to select device"))
        }
        return devices
    }

    private func refreshDiscoveredDevices() {
        let devices = getAggregatedDevices(); log("Discovered \(devices.count) device(s)")
        flutterApi.onDevicesChanged(devices: devices) { _ in }
    }

    private nonisolated func log(_ message: String) {
        os_log("[CastingHostApi] %{public}@", log: .default, type: .debug, message)
    }
}
